README.md
# Description

The __ihs__ cookbook uses IBM's Install Manager to install IBM HTTP Server.

## Basic Use

The __ihs__ cookbook only requires an IBM repository that contains IHS. The attribute [:ihs][:install][:repositoryLocation] __must__ be set to he location of your repository.

If you do not know where to find a repository, you can start at this page: http://pic.dhe.ibm.com/infocenter/wasinfo/v8r5/index.jsp?topic=%2Fcom.ibm.websphere.installation.express.doc%2Fae%2Fcins_repositories.html


You may also need to set [:ihs][:install][:secureStorageFile] to the location of your secure storage file in order to access a repository, and if your secure storage file is password protected you will also need to point [:ihs][:install][:masterPasswordFile] to your master password file.

This cookbook also supports keyfiles, however keyfiles are depreciated and we do not recommend their use.

# Requirements

## Platform:

* Linux

## Cookbooks:

* iim

# Attributes

* `node[:ihs][:paths][:shared]` - The path to the shared directory for IBM products. Defaults to `"/opt/IBM/IMShared"`.
* `node[:ihs][:paths][:install]` - The path where IHS should be installed. Defaults to `"/opt/IBM/HTTPServer"`.
* `node[:ihs][:paths][:eclipse]` - The path where IHS internal eclipse tools should be installed. Defaults to `"node[:ihs][:paths][:install]"`.
* `node[:ihs][:paths][:plugins]` - The path where Web Server Plug-ins for IBM WebSphere Application Server should be installed. Defaults to `"/opt/IBM/WebSphere/Plugins"`.
* `node[:ihs][:settings][:arch]` - The archetecture being installed on. Defaults to `"x86"`.
* `node[:ihs][:settings][:port]` - The port for IHS. Defaults to `"80"`.
* `node[:ihs][:install][":bits"]` - Install a 64 or 32 bit version? Defaults to checking automatically. Defaults to `""`.
* `node[:ihs][:install][:repositoryLocation]` - The location of the repository to download IHS from, this attribute is REQUIRED. Defaults to `"nil"`.
* `node[:ihs][:install][:masterPasswordFile]` - The location of the master password file IIM should use to access the secure storage file, this attribute is optional. Defaults to `"nil"`.
* `node[:ihs][:install][:secureStorageFile]` - The location of the secure storage file IIM should use to access the repoistory, this attribute is optional. Defaults to `"nil"`.
* `node[:Robin_Config][:backEnd][:java][:toolsPath]` - The location of tools.jar on the liberty servers, this is needed to generate plugin-cfg.xml. As an alternative you can put a copy of tools.jar into the DMZIHSLiberty/files/defualt directory and leave this peramater as nil. Defaults to `"nil"`.
* `node[:Robin_Config]["backEnd"]["findGateway"]["pattern"]` - How the liberty servers should search for the front end IBM HTTP Server. This must be set on the back end. Examples of valid values for pattern: "hostname:<name>", "name:<chef-node-name>", "role:<chef-role>". If multiple results are returned this recipe will send the plugin-cfg.xml files to all of them. For more details see: https://docs.getchef.com/essentials_search.html#partial-search. Defaults to `"nil"`.
* `node[:Robin_Config]["backEnd"]["findGateway"]["networkInterface"]` -  Defaults to `"eth0"`.
* `node[:Robin_Config][:backEnd][:findGateway][:overrideIP]` - If this is set Chef will use the IPs provided instead of searching for the servers. This value takes an array of ips. Defaults to `"nil"`.
* `node[:Robin_Config][:backEnd][:SSH][:username]` - The back end liberty servers require the ability to SSH to the front end. !Warning, this recipy can use SSH usernames and passwords, this is NOT recommended. Defaults to `"nil"`.
* `node[:Robin_Config][:backEnd][:SSH][:password]` -  Defaults to `"nil"`.
* `node[:Robin_Config][:backEnd][:SSH][:supressUnkownHostsWarning]` -  Defaults to `"true"`.
* `node[:Robin_Config][:backEnd][:SSH][:override]` -  Defaults to `"nil # if not nil, a custom SSH string will be used."`.
* `node[:Robin_Config][:frontEnd][:IHS][:baseDir]` - Where to find the front end IBM HTTP Server. If left to nil then this will fall back to [:ihs][:paths][:install]. If you wish to run configihs One of the two variables __must__ be set. Defaults to `"nil"`.
* `node[:Robin_Config][:frontEnd][:IHS][:pluginBaseDir]` - Where to find the Web Server Plug-ins for IBM WebSphere Application Server V8.5. If left to nil then this will fall back to [:ihs][:paths][:plugins]. If you wish to run configihs one of the two variables __must__ be set. Defaults to `"nil"`.
* `node[:Robin_Config][:frontEnd][:IHS][:install][:responseFile]` - The reponse file and optionally a secure storage file for install manager to install IHS. Defaults to `"nil"`.
* `node[:Robin_Config][:frontEnd][:IHS][:install][:secureStorageFile]` -  Defaults to `"nil"`.
* `node[:Robin_Config][:frontEnd][:plugincfg][:logPath]` - Peramaters to override settings in the final config-xml these should be set on the front end server. Defaults to `"nil"`.
* `node[:Robin_Config][:frontEnd][:plugincfg][:hostname]` -  Defaults to `"*"`.
* `node[:Robin_Config][:frontEnd][:plugincfg][:ports]` -  Defaults to `"[ ... ]"`.

# Recipes

* [ihs::default](#ihsdefault)
* ihs::robin_collect
* ihs::robin_config

## ihs::default

This recipe calls IIM and uses it to install IHS

# License and Maintainer

Maintainer:: IBM (<>)

License:: Apache 2.0


CHANGELOG.md
ihs CHANGELOG
=============

This file is used to list changes made in each version of the ihs cookbook.

0.1.0
-----
- [Benjamin Confino] - Initial release of ihs




spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: ihs
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'ihs::default' do
  context 'When all attributes are default, on an unspecified platform' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/cookbooks/robin_test/metadata.rb
name             'robin_test'
license          'Apache 2.0'
description      'Test cookbook for ci.chef.ihs.config cookbook'
version          '0.1.0'

depends "apt"
depends "java"
depends "wlp"
depends "ihs"


test/cookbooks/robin_test/recipes/install.rb
# Cookbook Name:: ci.chef.ihs.config_start
# Attributes:: default
#
# (C) Copyright IBM Corporation 2013.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#Step one, install prerequisites: 


include_recipe "apt"
include_recipe "java"

#Step two, install liberty for the backend and two servers

include_recipe "wlp"



wlp_server "testone" do
  config ({
            "featureManager" => {
              "feature" => [ "jsp-2.2" ]
            },
            "httpEndpoint" => {
              "id" => "defaultHttpEndpoint",
              "host" => "*",
              "httpPort" => "9080",
              "httpsPort" => "9443"
            }
          })
  action :create
end

directory "#{node['wlp']['base_dir']}/wlp/usr/servers/testone/dropins/test.war" do
   recursive true
end

cookbook_file "#{node['wlp']['base_dir']}/wlp/usr/servers/testone/dropins/test.war/testone.jsp" do
  backup false
  source "testone.jsp"
  action :create_if_missing
end

wlp_server "testone" do
  action :start
end



wlp_server "testtwo" do
  config ({
            "featureManager" => {
              "feature" => [ "jsp-2.2" ]
            },
            "httpEndpoint" => {
              "id" => "defaultHttpEndpoint",
              "host" => "*",
              "httpPort" => "9081",
              "httpsPort" => "9444"
            }
          })
  action :create
end

directory "#{node['wlp']['base_dir']}/wlp/usr/servers/testtwo/dropins/test.war" do
   recursive true
end

cookbook_file "#{node['wlp']['base_dir']}/wlp/usr/servers/testtwo/dropins/test.war/testtwo.jsp" do
  backup false
  source "testtwo.jsp"
  action :create_if_missing
end

wlp_server "testtwo" do
  action :start
end


#Step three, create a frontend server

include_recipe "ihs" # note, also installs IM

#Step four, "transfer" the config to the server and then apply it

include_recipe "ihs::robin_collect"
include_recipe "ihs::robin_config"


#Step five, download some pages to /tmp so that you can test for their existance later. 

#in my experience it suceeds twice then fails twice - since I have been bad and made round robin urls that only work on one server
bash "download from server one through the proxy" do
  user "root"
  ignore_failure true
  cwd "/tmp"
  code <<-EOH
  wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp || wget localhost/test/testone.jsp
  EOH
end


bash "download from server two through the proxy" do
  user "root"
  ignore_failure true
  cwd "/tmp"
  code <<-EOH
  wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp || wget localhost/test/testtwo.jsp
  EOH
end 


libraries/merge_xml.rb
module MergeXML

class UriGroup
  attr_reader :oldName, :childURIs, :name, :included

  def initialize(node)
    @included = false 

    @oldName = node["Name"]
 
    @childURIs = Array.new
    uris = node.xpath("./Uri")
    uris.each do |u|
      uri = Uri.new(u)
      @childURIs << uri
    end
    

    tmp = ::Array.new
    @childURIs.each do |u|
      tmp << u.name.gsub(/[^0-9a-zA-Z]/i, '_')
    end
    @name = tmp.sort!.join(' ')

    tmp2 = node.attributes() 
    tmp2.delete("Name")
    @attrs = tmp2
  end

  def is_included?
    return @included
  end

  def include!
    @included = true
  end

  def is_different?(otherUriGroup)

    if not @childURIs.length == otherUriGroup.childURIs.length then return true end

    childrenOne = ::Array.new
    childrenTwo = ::Array.new

    @childURIs.each do |c| 
      childrenOne << c.name
    end
    childrenOne.sort!
 
    otherUriGroup.childURIs.each do |c| 
      childrenTwo << c.name
    end
    childrenTwo.sort!

    if childrenOne.eql?(childrenTwo) then return false else return true end
    
  end
  
  def to_xml_string 
    xmlAttrs = @attrs.map{|k,v| "#{k}=\"#{v}\""}
    xmlAttrs2 = xmlAttrs.join(' ')


    childNodes = ::Array.new
    @childURIs.each{|child| childNodes << child.to_xml_string}
    childNodes2 = childNodes.join(' ')
    xml = "<UriGroup Name=\"#{@name}\" #{xmlAttrs2} > \n"+
          "#{childNodes2} \n"+
          "</UriGroup>"
    return xml
  end
end

class Uri
  attr_reader :name

  def initialize(node)
    @name = node["Name"]
    tmp = node.attributes() 
    tmp.delete("Name")
    @attrs = tmp
  end
  
  def to_xml_string 
    xmlAttrs = @attrs.map{|k,v| "#{k}=\"#{v}\""}
    xmlAttrs2 = xmlAttrs.join(' ')
    xml = "<Uri Name=\"#{@name}\" #{xmlAttrs2} />"
    return xml
  end
end

class Route
   attr_reader :oldServerCluster, :oldUriGroup, :oldVirtualHostGroup
   attr_accessor :serverCluster, :uriGroup, :virtualHostGroup

   def initialize(node, serverCluster, uriGroup)
     @oldServerCluster = node["ServerCluster"]
     @oldUriGroup = node["UriGroup"]
     @oldVirtualHostGroup = node["VirtualHostGroup"]

     @serverCluster = serverCluster
     @uriGroup = uriGroup
     @virtualHostGroup = "default_host"
   end

   def to_xml_string 
      xml = "<Route ServerCluster=\"#{@serverCluster.name}\" UriGroup=\"#{@uriGroup.name}\" VirtualHostGroup=\"#{@virtualHostGroup}\"  />"
      return xml
   end
end

class ServerCluster
  attr_reader :name, :oldName, :included, :apps
  attr_accessor :servers

  def initialize(node)
    @included = false
    @apps = ::Set.new
 
    @oldName = node["Name"]
    tmp = node.attributes() 
    tmp.delete("Name")
    @attrs = tmp  
    
    serverList = Array.new
    servers = node.xpath("./Server")
    servers.each do |s|
       ser = Server.new(s)
       serverList << ser
    end
    @servers = serverList   

    self.update_name!
  end

  def update_name!
    servers = ::Array.new
    @servers.each do |s|
      servers << s.name
    end
    servers.sort!

    apps = ::Array.new 
    @apps.each do |a|
      apps << a.gsub(/[^0-9a-zA-Z]/i, '')
    end
    apps.sort!

    @name = servers.join('_') + "_" + apps.join('_')
  end

  def is_different?(otherServerCluster)
    myAppNames = ::Set.new
    otherAppNames = ::Set.new

    @apps.each do |app|
      myAppNames << app
    end
   
    otherServerCluster.apps.each do |app|
      otherAppNames << app
    end


    if myAppNames == otherAppNames
      return false
    else
      return true
    end
  end

  def add_app!(app)
    @apps << app
    update_name!
  end

  def merge!(otherServerCluster)
  
    otherServerCluster.servers.each do |newServer|
      found = false
      @servers.each do |inServer|
        if newServer.name == inServer.name then found = true end
      end
      if not found then servers << newServer end
    end
    self.update_name!
  end

  def is_included?
    return @included
  end

  def include!
    @included = true
  end

  def to_xml_string 
    xmlAttrs = @attrs.map{|k,v| "#{k}=\"#{v}\""}
    xmlAttrs2 = xmlAttrs.join(' ')
    childNodes = ::Array.new 
    childNodes << ""
    @servers.each{|child| childNodes << child.to_xml_string}
    childNodes2 = childNodes.join(' ')
    xml = "<ServerCluster Name=\"#{@name}\" #{xmlAttrs2}> \n"+
          "#{childNodes2}"+
          "\n </ServerCluster>"
    return xml
  end
end

class Server
  attr_reader :name

  def initialize(node)
    @oldName = node["Name"]
    
    tmp = node.attributes() 
    tmp.delete("Name")
    @attrs = tmp      

    tmp2 = ::Array.new
    trans = node.xpath("./Transport")
    trans.each do |t|
      tmp2 << t["Hostname"] + ":" + t["Port"]
    end
    @name = tmp2.sort!.join(' ')
    @children = trans 
  end

  def to_xml_string 
  xmlAttrs = @attrs.map{|k,v| "#{k}=\"#{v}\""}
  xmlAttrs2 = xmlAttrs.join(' ')
  
  childNodes = ::Array.new
  childNodes << ""
  @children.each do |c|
    childNodes << c.to_s
  end

  xml = "<Server Name=\"#{@name}\" #{xmlAttrs2}> \n"+
        "#{childNodes.join(' ')}"+
        "</Server>"
    return xml
  end
end

def self.mergeXML(filepath1, filepath2, outputFile, chefNode)
    require 'nokogiri'
    require 'tempfile'
    file1 = File.open(filepath1)
    docOne = Nokogiri::XML(file1)
    file1.close 

    file2 = File.open(filepath2)
    docTwo = Nokogiri::XML(file2)
    file2.close 


   #Build the server objects
   
 
   serverClustersOne = ::Array.new
   serverClusters = docOne.xpath("/Config/ServerCluster")
   serverClusters.each do |sc|
     newSC = ServerCluster.new(sc)
     serverClustersOne << newSC
   end
   
   serverClustersTwo = ::Array.new
   serverClusters = docTwo.xpath("/Config/ServerCluster")
   serverClusters.each do |sc|
     newSC = ServerCluster.new(sc)
     serverClustersTwo << newSC
   end


   #Build the URI Group objects

   uriGroupsOne = ::Array.new
   uriGroups = docOne.xpath("/Config/UriGroup")
   uriGroups.each do |ug|
     newUG = UriGroup.new(ug)
     uriGroupsOne << newUG unless newUG.childURIs.empty?
   end
   
   uriGroupsTwo = ::Array.new
   uriGroups = docTwo.xpath("/Config/UriGroup")
   uriGroups.each do |ug|
     newUG = UriGroup.new(ug)
     uriGroupsTwo << newUG unless newUG.childURIs.empty?
   end

   #build the routes
   routesOne = ::Array.new
   routesDocOne = docOne.xpath("/Config/Route")
   routesDocOne.each do |route|

     newRouteUriGroup = nil
     newRouteserverCluster = nil
     
     uriGroupsOne.each do |ug|
       done = false
       tmp = route["UriGroup"]
       if tmp == ug.oldName and not done
         newRouteUriGroup = ug
         done = true
       end
     end
     
     serverClustersOne.each do |sc|
       done = false
       tmp = route["ServerCluster"]
       if tmp == sc.oldName and not done
         newRouteserverCluster = sc
         done = true
       end
     end
   
     if not newRouteserverCluster.nil? and not newRouteUriGroup.nil?
       newRoute = Route.new(route, newRouteserverCluster, newRouteUriGroup)
       routesOne << newRoute
     end
   end



   routesTwo = ::Array.new
   routesDocTwo = docTwo.xpath("/Config/Route")
   routesDocTwo.each do |route|

     newRouteUriGroup = nil
     newRouteserverCluster = nil
    
     uriGroupsTwo.each do |ug|
       done = false
       tmp = route["UriGroup"]
       if tmp == ug.oldName and not done
         newRouteUriGroup = ug
         done = true
       end
     end

     serverClustersTwo.each do |sc|
       done = false
       tmp = route["ServerCluster"]
       if tmp == sc.oldName and not done
         newRouteserverCluster = sc
         done = true
       end
     end
     
     if not newRouteserverCluster.nil? and not newRouteUriGroup.nil?
       newRoute = Route.new(route, newRouteserverCluster, newRouteUriGroup)
       routesTwo << newRoute 
     end
   end

   #ServerClusters will be merged if they have the same apps, so the routes shall be used to record apps in the server object.

   routesOne.each do |route|
     route.uriGroup.childURIs.each do |uri|
       route.serverCluster.add_app!(uri.name)
     end
   end

   routesTwo.each do |route|
     route.uriGroup.childURIs.each do |uri|
       route.serverCluster.add_app!(uri.name)
     end
   end
 

   ################# 
   #Now start the merging. 

   #Start with the servers
   #Server clusters will be merged if the names are the same, and the name is a sorted combination of all the hostnames in the cluster
   #As always the first file shall dominate. 
   serverClustersCombined = ::Array.new
   

   serverClustersOne.each do |newSC|
     found = false

     if serverClustersCombined.empty?
       serverClustersCombined << newSC
       newSC.include!
       next
     end

     serverClustersCombined.each do |inSC|
       if not newSC.is_different?(inSC) 
         found = true
         inSC.merge!(newSC)
       end 
     end
     if not found
      serverClustersCombined << newSC 
      newSC.include! 
     end
   end
   
   serverClustersTwo.each do |newSC|
     found = false

     if serverClustersCombined.empty?
       serverClustersCombined << newSC
       newSC.include!
       next
     end

     serverClustersCombined.each do |inSC|
       if not newSC.is_different?(inSC) 
         found = true
         inSC.merge!(newSC)
       end
     end
     if not found
      serverClustersCombined << newSC 
      newSC.include! 
     end
   end
   #######

   #next do the URIGroups
   #URIGroups will be merged if the names of all child URI's are the same.
   #As always the first file shall dominate. 

   uriGroupsCombined = ::Array.new
   
   uriGroupsOne.each do |newURI|
     if uriGroupsCombined.empty?
       uriGroupsCombined << newURI
       newURI.include!
       next
     end

     found = false
     uriGroupsCombined.each do |inURI|
       if not newURI.is_different?(inURI)
         found = true
       end
     end

     if not found     
       uriGroupsCombined << newURI
       newURI.include!
     end
   end

   uriGroupsTwo.each do |newURI|
     if uriGroupsCombined.empty?
       uriGroupsCombined << newURI
       newURI.include!
       next
     end

     found = false
     uriGroupsCombined.each do |inURI|
       if not newURI.is_different?(inURI)
         found = true
       end
     end

     if not found     
       uriGroupsCombined << newURI
       newURI.include!
     end
   end

   

   ##############
   # Merge the route objects. 
   # Rout objects will be the same if all three peramaters are the same. 

   routesCombined = ::Array.new

   routesOne.each do |newR|
     found = false
 
     if routesCombined.empty?
       routesCombined << newR
       next
     end

     routesCombined.each do |inR|
       if newR.uriGroup.name == inR.uriGroup.name and newR.serverCluster.name == inR.serverCluster.name
         found = true 
       end
     end
     if not found and newR.uriGroup.included and newR.serverCluster.included
       routesCombined << newR 
     end
   end

   routesTwo.each do |newR|
     found = false

     if routesCombined.empty?
       routesCombined << newR
       next
     end

     routesCombined.each do |inR|
       if newR.uriGroup.name == inR.uriGroup.name and newR.serverCluster.name == inR.serverCluster.name
         found = true 
       end
     end
     if not found and newR.uriGroup.included and newR.serverCluster.included
       routesCombined << newR 
     end
   end


############### Build the XML


    builder = Nokogiri::XML::Builder.new do
      Config {
      }
    end

    docCombined = Nokogiri::XML(builder.to_xml) do |config|
      config.default_xml.noblanks
    end

    rootAttr = docOne.xpath("/Config").first.attributes
    xmlNode = docCombined.at("Config")
    rootAttr.each do |name, value| 
      xmlNode[name] = value
    end

    #Copy the <Log> element from any of the two merging files, and modify the Path/Name, if required. 
    
    logAttr = docOne.xpath("/Config/Log").first
    docCombined.at("Config").add_child(logAttr)

    if not chefNode[:logPath].nil?
      docCombined.xpath("/Config/Log").first["Name"] = chefNode[:logPath]
    end

    #Copy properties like esiEnable and esiMaxCacheSize from any of the two merging files.

    propertyList = docOne.xpath("/Config/Property")
    propertyList.each do |prop|
      docCombined.at("Config").add_child(prop)
    end    

    #For this cookbook we will have all the servers in one virtual host group. In practice this means that the DMZ can only serve one domain. 

    ports = chefNode[:ports]
 

    virtualHostFragment = '<VirtualHostGroup Name="default_host"> '
    ports.each do |port|
      vhost = '<VirtualHost Name="' + chefNode[:hostname]+':'+port+'"/>'
      virtualHostFragment << vhost
    end
    

    virtualHostFragment << '</VirtualHostGroup>' 
    
    newNodes = ::Array.new
 
    newNodes << virtualHostFragment    

    serverClustersCombined.each do |newNode| 
      newNodes <<  newNode.to_xml_string
    end

    uriGroupsCombined.each do |newNode| 
      newNodes << newNode.to_xml_string
    end

    routesCombined.each do |newNode|
      newNodes <<  newNode.to_xml_string
    end

    newNodes.each do |newNode|
       docCombined.at("Config").add_child(newNode)
    end

   tmp = Tempfile.new('mergedFile') #I'm sure there's a better way to pretty print XML, but hanven't been able to find it.

    
   File.open(tmp,'w') do |file|
     file.truncate(0) 
     file.puts(docCombined.to_xml(:indent => 2))
   end 

   tempXML = ""
   File.open(tmp,'r') do |file|
     tempXML = file.read
   end 

   finalXML = Nokogiri.XML(tempXML) do |config|
      config.default_xml.noblanks
   end

   
   File.open(outputFile,'w') do |file|
     file.truncate(0) 
     file.puts(finalXML.to_xml(:indent => 2))
   end 

end

  

end


recipes/robin_config.rb
#
# Cookbook Name:: Robin_Config
# Recipe:: default
#
# Copyright 2014, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#
require "open3"
require 'tempfile'
require 'fileutils'
chef_gem 'nokogiri'

class Chef::Recipe
  include MergeXML
end


if not node[:Robin_Config][:frontEnd][:IHS][:baseDir].nil? 
  baseDir = node[:Robin_Config][:frontEnd][:IHS][:baseDir]
elsif !node[:ihs].nil? and !node[:ihs][:paths].nil? and !node[:ihs][:paths][:install].nil? #As part of a different cookbook this might not exist at all, checking extensively to prevent errors 
  baseDir = node[:ihs][:paths][:install]
else
  raise "The base directory of IHS was not found. Please add it to the node.conf or a suitable alternative"
end


if not node[:Robin_Config][:frontEnd][:IHS][:webSpherePluginsDir].nil? 
  webSpherePluginsDir = node[:Robin_Config][:frontEnd][:IHS][:webSpherePluginsDir]
elsif !node[:ihs].nil? and !node[:ihs][:paths].nil? and !node[:ihs][:paths][:plugins].nil? #As part of a different cookbook this might not exist at all, checking extensively to prevent errors 
  webSpherePluginsDir = node[:ihs][:paths][:plugins]
else
  raise "The base directory of WebSphere plugins directory was not found. Please add it to the node.conf or a suitable alternative"
end


bits = %x(uname -m | grep -o "^[^_]*" | sed -e 's/i6/i3/g')

if bits.start_with?('x') #as in x86_64
  bits = '64bits/'
elsif bits.start_with?('i') #as in i386 or i686
  bits = '' #this should be empty. 
end

binDir = "#{baseDir}/bin"
confDir = "#{baseDir}/conf"
plugincfg = "#{confDir}/plugin-cfg.xml"
mergedFilePrime = Tempfile.new('mergedFile')
mergedFilePrime.write(".")#force file creation, it will be overwritten later. 
mergedFile = mergedFilePrime.path
mergedFilePrime.close()

ruby_block "create plugin-cfg.xml" do 
  block do
   unmergedFiles = []
   

   Dir.chdir("/tmp/plgin-cfgs")
   Dir.glob('**/*.xml').each do |item|
     next if item == '.' or item == '..'
       unmergedFiles << item
     end

   if unmergedFiles.length == 0 
     raise "No plgin-cfg.xml files were found."
   elsif unmergedFiles.length == 1
     %x(cp #{unmergedFiles[0]} #{mergedFile})
   else

     MergeXML.mergeXML(unmergedFiles[0], unmergedFiles[1], mergedFile, node[:Robin_Config][:frontEnd][:plugincfg])

     #next merge any additional files into the previous merged file. 
     i = 2
     while i <  unmergedFiles.length do
      MergeXML.mergeXML(mergedFile, unmergedFiles[i], mergedFile, node[:Robin_Config][:frontEnd][:plugincfg])
      i += 1
     end
   end
  end
end

bash "create empty plugin-cfg" do # Later on I compare to see if plugincfg needs to be updated, so I need something to compare against. 
  code <<-EOH
  cp -p #{confDir}/httpd.conf #{plugincfg}
  EOH
  not_if {::File.exists?(plugincfg)}
end


ruby_block "load websphere plugin" do
  block do

    lines = ::IO.readlines("#{confDir}/httpd.conf")
    index = lines.find_index { |line| line.start_with?("LoadModule")}
    lines.insert(index, "LoadModule was_ap22_module #{webSpherePluginsDir}/bin/#{bits}mod_was_ap22_http.so")

    ::File.open("#{confDir}/httpd.conf", 'w') do |file|
      file.puts lines
    end
  end
  not_if {::File.open("#{confDir}/httpd.conf").lines.any?{|line| line.include?("LoadModule was_ap22_module")}}
end


updateRequired = false

if ::File.exists?(plugincfg)
  updateRequired = ::FileUtils.compare_file(plugincfg, mergedFile) #returns true if the files are identical
  updateRequired = !updateRequired #so reverse it 
else
  updateRequired = true
end


bash "stop HTTP Server" do
  cwd binDir 
  code <<-EOH
  ./apachectl stop
  ./adminctl stop
  EOH
  only_if {updateRequired}
end

#bash "stop HTTP Admin" do #by default HTTP Admin is not configured and not run
#  cwd binDir 
#  code <<-EOH
#  ./adminctl stop
#  EOH
#  only_if {updateRequired}
#  ignore_failure true# TODO It is valid to run without adminctl at all; can this be handled more elegantly than trying and ignoring failures?
#end

bash "append httpd.conf" do
  cwd confDir
  code <<-EOH
  sed -i '/WebSpherePluginConfig/d' httpd.conf
  echo 'WebSpherePluginConfig #{confDir}/plugin-cfg.xml' >> httpd.conf
  EOH
  only_if {updateRequired}
end

bash "place new plugin-cfg" do
  cwd confDir
  code <<-EOH 
  cp #{mergedFile} #{plugincfg}
  echo #{mergedFile} >> /tmp/cheflog
  echo #{plugincfg}  >> /tmp/cheflog
  EOH
  only_if {updateRequired}
end
 

bash "start HTTP Server" do
  cwd binDir
  code <<-EOH
  ./apachectl start
  EOH
  only_if {updateRequired}
end

#bash "start HTTP Admin" do
#  cwd binDir
#  code <<-EOH
#  ./adminctl start
#  EOH
#  only_if {updateRequired}
#  ignore_failure true# TODO It is valid to run without adminctl at all; can this be handled more elegantly than trying and ignoring failures?
#end





recipes/default.rb
=begin
#<
This recipe calls IIM and uses it to install IHS
#>
=end

# Cookbook Name:: ihs
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
require 'tempfile'

#TODO politely exit if node[:ihs][:install][:repositoryLocation] is nil

include_recipe "iim"


bits = %x(uname -m | grep -o "^[^_]*" | sed -e 's/i6/i3/g')
if bits.start_with?('x') #as in x86_64
  node.default[:ihs][:install][":bits"] = '64'
elsif bits.start_with?('i') #as in i386 or i686
  node.default[:ihs][:install][":bits"] = '32'
end

responseFile = Tempfile.new('response.xml')
responseFilePath = responseFile.path

directory "/opt/IBM" do
  group node[:im][:group]
  owner node[:im][:user]
  mode "0755"
  recursive true
end

template responseFilePath do
  source "install_response_file.xml.erb"
  action :create
  group node[:im][:group]
  owner node[:im][:user]
  #all variables are set in the erb itself.
end

# copy secureStorageFile and masterPasswordFile so we can connect to the repo
# %w(secureStorageFile masterPasswordFile).each do |file|
  #c ookbook_file "/var/tmp/#{file}" do
  #
  #  source "#{file}"
  #  owner node['im']['user']
  #  group node['im']['group']
  #  mode 00644
  #  action :create
  # end
# end

# iim_install 'ihs' do
iim_response_file_install 'ihs' do
  response_file responseFilePath
  # all these default to null, and depending on your repositories security settings they might be able to remain that way.
  master_password_file node[:ihs][:install][:masterPasswordFile]
  secure_storage_file node[:ihs][:install][:secureStorageFile]
  action :install
end

template "/etc/init.d/ibm-http" do
  source "IBM-http.sh.erb"
  group "root"
  owner "root"
  mode "0755"
end

service "ibm-http" do
  supports :restart => true, :reload => true
  action [ :enable ]
  provider Chef::Provide::Service:AixInit if platform?('aix')
end
